src/BuiltInTools/BrowserRefresh/ResponseStreamWrapper.cs

```diff
@@ -244,29 +109,32 @@ private void OnWrite()
                 // Since we're changing the markup content, reset the content-length
                 response.Headers.ContentLength = null;
 
+                _scriptInjectingStream = new ScriptInjectingStream(_baseStream);
+
+                // By default, write directly to the script injection stream.
+                // We may change the base stream below if we detect that the response
+                // is compressed.
+                _baseStream = _scriptInjectingStream;
+
                 // Check if the response has gzip Content-Encoding
                 if (response.Headers.TryGetValue(HeaderNames.ContentEncoding, out var contentEncodingValues))
                 {
                     var contentEncoding = contentEncodingValues.FirstOrDefault();
                     if (string.Equals(contentEncoding, "gzip", StringComparison.OrdinalIgnoreCase))
                     {
-                        _isGzipEncoded = true;
                         // Remove the Content-Encoding header since we'll be serving uncompressed content
                         response.Headers.Remove(HeaderNames.ContentEncoding);
-                        InitializeBuffers();
+
+                        _pipe = new Pipe();
+                        var gzipStream = new GZipStream(_pipe.Reader.AsStream(leaveOpen: true), CompressionMode.Decompress, leaveOpen: true);
+
+                        _gzipCopyTask = gzipStream.CopyToAsync(_scriptInjectingStream);
+                        _baseStream = _pipe.Writer.AsStream(leaveOpen: true);
```

Comment:
I believe I tried this, and I think I discarded it because I wanted to try and "stream" the response as much as possible. With this approach we are essentially buffering the entire response, aren't we?

That said, as long as it works, I'm happy. if the test that pushes the Gzipped payload 1 bit at a time works that works.

Reply:
I don't mind in any case if we buffer the response, as long as things work.

Reply:
I think the key was that I was able to see how many uncommited bytes I had on the writer, but I didn't have a way to see how many unread bytes I had on the pipestream reader, which is what I used to call Read on the gzip stream.